extensions/wrapper/wrapper/src/main/java/de/sovity/edc/ext/wrapper/api/ui/pages/data_offer/DataOfferPageApiService.java

```diff
@@ -66,45 +67,90 @@ private boolean isIdAvailable(DSLContext dsl, Table<?> table, TableField<?, Stri
     }
 
     public IdResponseDto createDataOffer(DSLContext dsl, DataOfferCreationRequest dataOfferCreationRequest) {
+
         val commonId = dataOfferCreationRequest.getUiAssetCreateRequest().getId();
 
-        val assetIdExists = checkIfAssetIdAvailable(dsl, commonId).isAvailable();
-        if (!assetIdExists) {
-            throw new InvalidRequestException("Asset with id %s already exists".formatted(commonId));
-        }
+        return switch (dataOfferCreationRequest.getPolicy()) {
+            case DONT_PUBLISH -> createButDontPublish(dsl, dataOfferCreationRequest, commonId);
+            case PUBLISH_UNRESTRICTED -> createAndPublishUnrestricted(dsl, dataOfferCreationRequest, commonId);
+            case PUBLISH_RESTRICTED -> createAndPublishRestricted(dsl, dataOfferCreationRequest, commonId);
+        };
+    }
 
-        val policyIdExists = checkIfPolicyIdAvailable(dsl, commonId).isAvailable();
-        if (!policyIdExists) {
-            throw new InvalidRequestException("Policy with id %s already exists".formatted(commonId));
-        }
+    private @NotNull IdResponseDto createAndPublishUnrestricted(DSLContext dsl, DataOfferCreationRequest dataOfferCreationRequest, String commonId) {
+        val assetId = commonId;
+        val contractDefinitionId = commonId;
+        val policyId = AlwaysTruePolicyConstants.POLICY_DEFINITION_ID;
 
-        val contractDefinitionIdExists = checkIfContractDefinitionIdAvailable(dsl, commonId).isAvailable();
-        if (!contractDefinitionIdExists) {
-            throw new InvalidRequestException("Contract definition with id %s already exists".formatted(commonId));
-        }
+        checkAssetIdAvailable(dsl, assetId);
+        checkContractDefinitionIdAvailable(dsl, contractDefinitionId);
+
+        assetApiService.createAsset(dataOfferCreationRequest.getUiAssetCreateRequest());
+
+        return createContractDefinition(assetId, policyId, contractDefinitionId);
+    }
+
+    private @NotNull IdResponseDto createAndPublishRestricted(DSLContext dsl, DataOfferCreationRequest dataOfferCreationRequest, String commonId) {
+        val assetId = commonId;
+        val policyId = commonId;
+        val contractDefinitionId = commonId;
+
+        checkAssetIdAvailable(dsl, assetId);
+        checkPolicyIdAvailable(dsl, policyId);
+        checkContractDefinitionIdAvailable(dsl, contractDefinitionId);
 
         assetApiService.createAsset(dataOfferCreationRequest.getUiAssetCreateRequest());
 
         val maybeNewPolicy = Optional.ofNullable(dataOfferCreationRequest.getUiPolicyExpression());
 
         maybeNewPolicy.ifPresent(
-            policy -> policyDefinitionApiService.createPolicyDefinitionV2(new PolicyDefinitionCreateDto(commonId, policy)));
+            policy -> policyDefinitionApiService.createPolicyDefinitionV2(new PolicyDefinitionCreateDto(policyId, policy)));
```

Comment:
You create a policy if the policy expression is not `null`. So if the expression is `null`, you don't create a policy but still pass the ID to the contract definition. What happens if the policy with this ID does not exist?

Reply:
Added exception when the policy definition is missing.